xrdp/xrdp.c

```diff
@@ -426,13 +426,29 @@ check_drop_privileges(struct xrdp_startup_params *startup_params)
     int rv = 1;
     const char *user = startup_params->runtime_user;
     const char *group = startup_params->runtime_group;
-
-    if (user[0] == '\0' && group[0] == '\0')
+    int entry_uid = g_getuid();
+    int entry_gid = g_getgid();
+
+    // Regard a user as privileged if either the UID or GID is zero
+    int privileged_user = (entry_uid == 0 || entry_gid == 0);
+    if (!privileged_user)
+    {
+        // xrdp daemon UID:GID is already set by whatever started it. We'll
+        // log this as it may be unexpected and continue. We don't have
+        // any privilege to override this.
+        LOG(LOG_LEVEL_INFO,
+            "xrdp is already running as an unprivileged user uid=%d gid=%d",
+            entry_uid, entry_gid);
+        rv = 0;
+    }
+    else if (user[0] == '\0' && group[0] == '\0')
     {
-        // Allow this for now
+        // We're running as a privileged user, and this isn't
+        // overridden in our own config
         LOG(LOG_LEVEL_ALWAYS,
-            "You are running xrdp as root. This is not safe.");
-        rv = 0;
+            "You are running xrdp as uid=%d gid=%d. This is not safe.",
```

Comment:
BTW, not directly related to the change, but "not safe" might sound too strong. It suggests being vulnerable or insecure (like not using TLS), whereas that’s not really the case. What about using "not recommended" instead?

Reply:
Agreed - that sound much better.